File: src/signals.zig
//! Signal handling for graceful shutdown.
//!
//! Handles SIGINT (Ctrl+C) and SIGTERM for clean termination.

const std = @import("std");

/// Global interrupt flag (use atomic operations)
var interrupted: std.atomic.Value(bool) = std.atomic.Value(bool).init(false);

/// Current issue being worked on (for cleanup messages)
var current_issue: ?[]const u8 = null;

/// Colors for terminal output
const Color = struct {
    const reset = "\x1b[0m";
    const yellow = "\x1b[1;33m";
};

/// Check if we've been interrupted
pub fn isInterrupted() bool {
    return interrupted.load(.acquire);
}

/// Set the current issue (for cleanup messages)
pub fn setCurrentIssue(issue: ?[]const u8) void {
    current_issue = issue;
}

/// Signal handler function
fn handleSignal(sig: i32) callconv(.c) void {
    _ = sig;
    interrupted.store(true, .release);

    // Print cleanup message to stderr (signal-safe via write syscall)
    const msg1 = "\n" ++ Color.yellow ++ "[INTERRUPTED]" ++ Color.reset ++ " Caught signal, cleaning up...\n";
    _ = std.posix.write(std.posix.STDERR_FILENO, msg1) catch {};

    if (current_issue) |issue| {
        const msg2 = Color.yellow ++ "[INTERRUPTED]" ++ Color.reset ++ " Issue ";
        _ = std.posix.write(std.posix.STDERR_FILENO, msg2) catch {};
        _ = std.posix.write(std.posix.STDERR_FILENO, issue) catch {};
        const msg3 = " was NOT completed\n";
        _ = std.posix.write(std.posix.STDERR_FILENO, msg3) catch {};

        const msg4 = Color.yellow ++ "[INTERRUPTED]" ++ Color.reset ++ " Issue status left as-is (check with: bd show ";
        _ = std.posix.write(std.posix.STDERR_FILENO, msg4) catch {};
        _ = std.posix.write(std.posix.STDERR_FILENO, issue) catch {};
        const msg5 = ")\n";
        _ = std.posix.write(std.posix.STDERR_FILENO, msg5) catch {};
    }
}

/// Install signal handlers for SIGINT and SIGTERM
pub fn install() void {
    const sigint_action = std.posix.Sigaction{
        .handler = .{ .handler = handleSignal },
        .mask = @as(std.posix.sigset_t, 0),
        .flags = 0,
    };

    std.posix.sigaction(std.posix.SIG.INT, &sigint_action, null);
    std.posix.sigaction(std.posix.SIG.TERM, &sigint_action, null);
}

/// Reset signal handlers to default
pub fn reset() void {
    const default_action = std.posix.Sigaction{
        .handler = .{ .handler = std.posix.SIG.DFL },
        .mask = @as(std.posix.sigset_t, 0),
        .flags = 0,
    };

    std.posix.sigaction(std.posix.SIG.INT, &default_action, null);
    std.posix.sigaction(std.posix.SIG.TERM, &default_action, null);
}

test "signal module compiles" {
    // Just verify the module compiles
    try std.testing.expect(!isInterrupted());
}

File: lib/noface/util/signals.ex
defmodule Noface.Util.Signals do
  @moduledoc """
  Signal handling for graceful shutdown.

  Handles SIGINT (Ctrl+C) and SIGTERM for clean orchestrator shutdown.
  """
  use GenServer
  require Logger

  defmodule State do
    @moduledoc false
    defstruct interrupted: false, current_issue: nil
  end

  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
  end

  @doc "Install signal handlers."
  @spec install() :: :ok
  def install do
    GenServer.call(__MODULE__, :install)
  end

  @doc "Reset interrupt flag."
  @spec reset() :: :ok
  def reset do
    GenServer.call(__MODULE__, :reset)
  end

  @doc "Check if interrupted."
  @spec interrupted?() :: boolean()
  def interrupted? do
    GenServer.call(__MODULE__, :interrupted?)
  end

  @doc "Set the current issue being worked on (for cleanup messages)."
  @spec set_current_issue(String.t() | nil) :: :ok
  def set_current_issue(issue_id) do
    GenServer.call(__MODULE__, {:set_current_issue, issue_id})
  end

  @doc "Get the current issue."
  @spec get_current_issue() :: String.t() | nil
  def get_current_issue do
    GenServer.call(__MODULE__, :get_current_issue)
  end

  # GenServer callbacks

  @impl true
  def init(_opts) do
    {:ok, %State{}}
  end

  @impl true
  def handle_call(:install, _from, state) do
    # In Elixir/Erlang, we typically handle shutdown via :init.stop/0
    # or by trapping exits. For SIGINT, the VM handles it.
    # We can use :os.set_signal/2 on supported platforms.
    try do
      :os.set_signal(:sigint, :handle)
      :os.set_signal(:sigterm, :handle)
    rescue
      _ -> :ok
    end

    {:reply, :ok, state}
  end

  def handle_call(:reset, _from, state) do
    {:reply, :ok, %{state | interrupted: false}}
  end

  def handle_call(:interrupted?, _from, state) do
    {:reply, state.interrupted, state}
  end

  def handle_call({:set_current_issue, issue_id}, _from, state) do
    {:reply, :ok, %{state | current_issue: issue_id}}
  end

  def handle_call(:get_current_issue, _from, state) do
    {:reply, state.current_issue, state}
  end

  @impl true
  def handle_info({:signal, :sigint}, state) do
    Logger.info("\n\e[1;33m[INTERRUPTED]\e[0m Caught signal, cleaning up...")

    if state.current_issue do
      Logger.info("\e[1;33m[INTERRUPTED]\e[0m Issue #{state.current_issue} was NOT completed")
      Logger.info("\e[1;33m[INTERRUPTED]\e[0m Issue status left as-is (check with: bd show #{state.current_issue})")
    end

    {:noreply, %{state | interrupted: true}}
  end

  def handle_info({:signal, :sigterm}, state) do
    Logger.info("\n\e[1;33m[INTERRUPTED]\e[0m Received SIGTERM, shutting down...")
    {:noreply, %{state | interrupted: true}}
  end

  def handle_info(_msg, state) do
    {:noreply, state}
  end
end
